### ISSUE TITLE
`index_time_` weirdness

### ISSUE THREAD

**Issue by xvandish** [livegrep/livegrep#331]
Hello,

`index_time`, as returned by the `Info` gRPC call, is returning either `13` or `14` as the timestamp. The `Codesearch` backend that serves the index and the `indexer` the creates the Index runs are both GKE pods, running on Docker images with the exact same Dockerfiles as what's in the main repo.

Things I've checked:
1. I port forwarded the Codesearch pod to my local computer, and spun up the Go webserver, and confirmed that `/debug/stats` reflects the weird `13` or `14` number by reporting that the index is very old, even though it was actually just a few minutes old when I checked.  For example, `{"index_age":1650085072}` (time in seconds since index was created).
2. In Google's Cloud Shell, I installed the `grpc_cli`, then in another tab of the shell, port forwarded the Codesearch pod's port to local. Then I ran `./grpc_cli call localhost:8080 CodeSearch.Info ''` and got back
```
trees......
index_time: 14
```
3. I `kubectl exec`'d both into the Codesearch backend pod and into an instance of a Cron based indexer and ran `date +%s` and both times the output looked fine. For example `1650134782`. And using that same value in Go gets the following output.
```
num = 1650134782
t := time.Unix(num, 0)
fmt.Printf("%v\n", t)

Output:
2022-04-16 18:46:22 +0000 UTC
```

So It's seems like the clock on both machines is ok, not what's going wrong in [these](https://github.com/livegrep/livegrep/blob/main/src/codesearch.cc#L511-L513) lines.
```
timeval now;
gettimeofday(&now, NULL);
index_timestamp_ = now.tv_sec;
```

Any ideas what's going on? I'll keep looking into this, but figured someone might have an idea. 

Thanks!

**Comment by xvandish** [comment]
Alright, more debugging -
TLDR - issue may be from using `stat.st_mtime` when loading the index

I added a few debug lines to [src/codesearch.cc](https://github.com/xvandish/livegrep/commit/09fe5dfc799d5731786c46df0fa41a9efecb5976)
```
index_timestamp_ = now.tv_sec;
+ fprintf(stdout, "now.tv_sec: %ld\n", now.tv_sec);
+ fprintf(stdout, "index_timestamp_: %lld\n", index_timestamp_);
```

and to [src/dump_load.cc](https://github.com/xvandish/livegrep/commit/008d91b7b9b15efaa0f65fc3ee574ad89be29a27)
```
+ fprintf(stderr, "about to try to load in index_timestamp_\n");
struct stat st;
assert(fstat(fd_, &st) == 0);
cs->index_timestamp_ = st.st_mtime;
+ fprintf(stderr, "loaded index with index_timestamp_: %lld\n", cs->index_timestamp_);
+ fprintf(stderr, "st.st_mtime: %ld\n", st.st_mtime);
```

And got the following output from an indexing run
```
...
2022-04-18T00:00:03.202135070Z now.tv_sec: 1650240000
2022-04-18T00:00:03.202188130Z index_timestamp_: 1650240000
```

And the following from the codesearch backend when it reloaded
```
...
2022-04-18T00:00:06.681011934Z Reload()
2022-04-18T00:00:45.602139844Z about to try to load in index_timestamp_
2022-04-18T00:00:45.602183244Z loaded index with index_timestamp_: 0
2022-04-18T00:00:45.602190734Z st.st_mtime: 0
2022-04-18T00:00:45.602567164Z Serving...
```

From [online stat man page](https://linux.die.net/man/2/stat) `st_mtime` is `time of last modification`. Using stat (in bash shell) gets the following:
```
ls -alh /data
-rw-r--r--    1 root root  80G Apr 18 00:00 index.idx

stat /data/index.idx
...
Modify: 2022-04-18 00:00:00.370146699 +0000 
Change: 2022-04-18 00:00:06.676140413 +0000
```

**Comment by LINKIWI** [comment]
Apart from the odd behavior described here (which unfortunately I don't have an explanation for), I've also observed a similar "issue" related to an inaccurate/inconsistent index timestamp. If there is a significant propagation delay between the time that the index is built offline and the time it's deployed to the cluster, the timestamp will be inaccurate since it's derived from a `stat(2)` of the index file.

I think it logically makes sense to encode the timestamp at which the index was built into the dump, so that it can be deterministically recovered at load time. Something like this:

```diff
diff --git a/src/codesearch.cc b/src/codesearch.cc
index 5adeab3..43087f9 100644
--- a/src/codesearch.cc
+++ b/src/codesearch.cc
@@ -505,13 +505,14 @@ void code_searcher::index_filenames() {
 void code_searcher::finalize() {
     assert(!finalized_);
     finalized_ = true;
-    index_filenames();
-    alloc_->finalize();
 
     timeval now;
     gettimeofday(&now, NULL);
     index_timestamp_ = now.tv_sec;
 
+    index_filenames();
+    alloc_->finalize();
+
     idx_data_chunks.inc(alloc_->end() - alloc_->begin());
     idx_content_chunks.inc(alloc_->end_content() - alloc_->begin_content());
 }
diff --git a/src/codesearch.h b/src/codesearch.h
index e2e95e9..cb32e65 100644
--- a/src/codesearch.h
+++ b/src/codesearch.h
@@ -163,6 +163,9 @@ public:
     int64_t index_timestamp() {
         return index_timestamp_;
     }
+    void set_index_timestamp(int64_t index_timestamp) {
+        index_timestamp_ = index_timestamp;
+    }
 
     class search_thread {
     public:
diff --git a/src/dump_load.cc b/src/dump_load.cc
index 4a650dd..d6b0dc5 100644
--- a/src/dump_load.cc
+++ b/src/dump_load.cc
@@ -77,6 +77,10 @@ protected:
         dump(&i);
     }
 
+    void dump_int64(uint64_t i) {
+        dump(&i);
+    }
+
     void dump_string(const string &str) {
         dump_int32(str.size());
         stream_.write(str.c_str(), str.size());
@@ -235,6 +239,10 @@ protected:
         return *(consume<uint32_t>());
     }
 
+    uint64_t load_int64() {
+        return *(consume<uint64_t>());
+    }
+
     string load_string() {
         uint32_t len = load_int32();
         uint8_t *buf = p_;
@@ -309,6 +317,9 @@ void codesearch_index::dump_metadata() {
     hdr_.name_off = stream_.tellp();
     dump_string(cs_->name());
 
+    hdr_.timestamp_off = stream_.tellp();
+    dump_int64(cs_->index_timestamp());
+
     map<const indexed_tree*, int> tree_ids;
 
     hdr_.refs_off = stream_.tellp();
@@ -440,6 +451,9 @@ load_allocator::load_allocator(code_searcher *cs, const string& path) {
 
     p_ = ptr<unsigned char>(hdr_->name_off);
     cs->set_name(load_string());
+
+    p_ = ptr<unsigned char>(hdr_->timestamp_off);
+    cs->set_index_timestamp((int64_t) load_int64());
 }
 
 
@@ -561,10 +575,6 @@ void load_allocator::load(code_searcher *cs) {
         cs->filename_positions_.push_back(make_pair(pos, sf));
     }
 
-    struct stat st;
-    assert(fstat(fd_, &st) == 0);
-    cs->index_timestamp_ = st.st_mtime;
-
     cs->finalized_ = true;
 }
 
diff --git a/src/dump_load.h b/src/dump_load.h
index 908882b..247b662 100644
--- a/src/dump_load.h
+++ b/src/dump_load.h
@@ -11,7 +11,7 @@
 #include <stdint.h>
 
 const uint32_t kIndexMagic   = 0xc0d35eac;
-const uint32_t kIndexVersion = 15;
+const uint32_t kIndexVersion = 16;
 
 // 16k is the page size on Apple M1 macs, which is the largest page
 // size of supported platforms. We use a consistent page size
@@ -24,6 +24,7 @@ struct index_header {
     uint32_t chunk_size;
 
     uint64_t name_off;
+    uint64_t timestamp_off;
 
     uint32_t ntrees;
     uint64_t refs_off;
```

**Comment by xvandish** [comment]
@LINKIWI nice, that's exactly where I was thinking of going next - agree that it makes sense to encode the creation time and should fix both of our weird cases. I'll try out your patch code (thanks btw!) and test it out on our deployment in ~30mins and report back.

I'm also going to try making some small dummy C++ and Go programs that try a few different variations of `stat` calls - (`fstat`, `os.Stat()` `system("some_bash_script_that_calls_stat.sh")` and see if any/what weird differences come up...

**Comment by xvandish** [comment]
@LINKIWI patch worked, `index_time` returned by `Info` looks good now.  

Indexer run:
```
...
== end metrics ==
now.tv_sec: 1650255576
index_timestamp_: 1650255576
```

and a gRPC call to the port-forwarded CodeSearch backend returns the same timestamp:
```
$ ./grpc_cli call localhost:8080 CodeSearch.Info ''                                                                                                        
trees......
index_time: 1650255576
Rpc succeeded with OK status
```

Thanks again!

**Comment by LINKIWI** [comment]
Nice, glad to hear it worked. I'll post this patch as a PR then.
